### PR TITLE
Clean up hash-to-field

### PR DESCRIPTION
**Pull request by burdges**
## Description

Fixes poor usage of Digest, unnecessary allocations, poor style, etc.  Nothing complex here.

Please merge #677 first.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (master)
- [ ] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: ff/src/fields/field_hashers/mod.rs
Comment (mmagician) [inline ff/src/fields/field_hashers/mod.rs:65]: nice

File: ff/src/fields/field_hashers/mod.rs
Comment (mmagician) [inline ff/src/fields/field_hashers/mod.rs:107]: good catch

File: ff/src/fields/mod.rs
Comment (mmagician) [inline ff/src/fields/mod.rs:228]: Ah I see, so this diff will be gone with merging #677

File: ff/src/fields/field_hashers/expander/mod.rs
Comment (mmagician) [inline ff/src/fields/field_hashers/expander/mod.rs:78]: CI complains that `hasher` is never used/read - I think it refers to its `(super)` visibility, it can probably stay private.

File: ff/src/fields/field_hashers/expander/mod.rs
Comment (mmagician) [inline ff/src/fields/field_hashers/expander/mod.rs:6]: So IIUC `Digest` is a supertrait of `FixedOutputReset` and we just don't need all of supertrait functionality?

File: ff/src/fields/field_hashers/expander/mod.rs
Comment (burdges) [reply ff/src/fields/field_hashers/expander/mod.rs:78]: We'll change this in a later PR but we've not changed this line yet, so the CI is wrong here.
https://github.com/arkworks-rs/algebra/pull/678/files#diff-ec6c8867ccebc8916382b7a4c0d910966930416987c40d56011dcf45fb88c872R54

This is a common incorrect warning.

**Comment by burdges** [comment]
Just fyi, I've also pulled out a branch which contains the larger subsequent changes to hash-to-field: https://github.com/burdges/algebra/commits/xof_h2f  It looks easier if it's merged with some curve changes though.

File: ff/src/fields/field_hashers/expander/mod.rs
Comment (Pratyush) [inline ff/src/fields/field_hashers/expander/mod.rs:30]: Let's do something like 
```suggestion
        let array = if dst.len() > MAX_DST_LENGTH {
            let mut long = H::default();
            long.update(&LONG_DST_PREFIX[..]);
            long.update(&dst);
            (long.finalize_fixed().as_ref()).try_into()
        } else {
            dst.try_into()
        };
        DST(array.unwrap())
```

File: ff/src/fields/field_hashers/expander/mod.rs
Comment (Pratyush) [inline ff/src/fields/field_hashers/expander/mod.rs:45]:
```suggestion
        let array = if dst.len() > MAX_DST_LENGTH {
            let mut long = H::default();
            long.update(&LONG_DST_PREFIX[..]);
            long.update(&dst);

            let mut new_dst = [0u8; MAX_DST_LENGTH];
            let new_dst = &mut new_dst[0..((2 * k + 7) >> 3)];
            long.finalize_xof_into(new_dst);
            new_dst.try_into()
        } else {
            dst.try_into()
        };
        DST(array.unwrap())
```

File: ff/src/fields/field_hashers/mod.rs
Comment (Pratyush) [inline ff/src/fields/field_hashers/mod.rs:20]: Let's update the comment as well.

File: ff/src/fields/field_hashers/mod.rs
Comment (Pratyush) [inline ff/src/fields/field_hashers/mod.rs:5]:
```suggestion
```

File: ff/src/fields/field_hashers/expander/mod.rs
Comment (Pratyush) [reply ff/src/fields/field_hashers/expander/mod.rs:78]: Hm given that we can't merge without it, what's an acceptable change for right now to get this PR merged?

Comment (Pratyush) [reply ff/src/fields/field_hashers/expander/mod.rs:78]: Maybe an `#[allow]` temporarily, that will be removed in a future PR?

Comment (burdges) [reply ff/src/fields/field_hashers/expander/mod.rs]: Oops!  It's fine as `PhantomData`.  This churns later too, but no issue here.  I'll rebase and push the fix.

Comment (burdges) [reply ff/src/fields/field_hashers/expander/mod.rs]: ok fixed now.

**Review by Pratyush — APPROVED** [review]
LGTM once tests are fixed!

File: ff/src/fields/field_hashers/expander/mod.rs
Comment (burdges) [reply ff/src/fields/field_hashers/expander/mod.rs]: Alright minor hiccup with tests not being run locally, but now fixed I think.

**Comment by burdges** [comment]
It's been a while so I've forgotten how to trigger the expander/tests.rs to be built locally.  lol

I figured this out one before.  I'd though it was cargo test form the repo root, but that doesn't seem to do it right now.  I should look at the CI maybe.

**Comment by burdges** [comment]
CI is happy now

**Comment by Pratyush** [comment]
There's a little conflict; you'll have to move the `digest` dependency to the root `Cargo.toml`

**Comment by burdges** [comment]
Alright fixed I think, thanks  :)

**Review by Pratyush — APPROVED** [review]

**Pull request merged**
